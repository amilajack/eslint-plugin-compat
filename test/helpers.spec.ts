import path from "path";
import {
  determineTargetsFromConfig,
  parseBrowsersListVersion,
} from "../src/helpers";
import multiEnvPackageJSON from "./multi-config.package.json";
import singleArrayEnvPackageJSON from "./single-array-config.package.json";
import singleVersionEnvPackageJSON from "./single-version-config.package.json";

describe("Versioning", () => {
  it("should support multi env config in browserslist package.json", () => {
    const config = determineTargetsFromConfig(
      ".",
      multiEnvPackageJSON.browsers
    );
    const result = parseBrowsersListVersion(config);
    expect(result).toMatchSnapshot();
  });

  it("should support resolving browserslist config in subdirectory", () => {
    // This should resolve the ./test/.browserslistrc config
    const relativeConfig = determineTargetsFromConfig(
      path.join(__dirname, ".browserslistrc")
    );
    const rootConfig = determineTargetsFromConfig(
      require.resolve("../package.json")
    );
    const relativeConfigVersions = parseBrowsersListVersion(relativeConfig);
    expect(relativeConfigVersions).toMatchSnapshot();
    const rootConfigVersions = parseBrowsersListVersion(rootConfig);
    expect(rootConfigVersions).toMatchSnapshot();

    expect(relativeConfigVersions).not.toEqual(rootConfigVersions);
  });

  it("should support single array config in browserslist package.json", () => {
    const config = determineTargetsFromConfig(
      ".",
      singleArrayEnvPackageJSON.browsers
    );
    const result = parseBrowsersListVersion(config);
    expect(result).toMatchSnapshot();
  });

  it("should support single version config in browserslist package.json", () => {
    const config = determineTargetsFromConfig(
      ".",
      singleVersionEnvPackageJSON.browsers
    );
    const result = parseBrowsersListVersion(config);
    expect(result).toMatchInlineSnapshot(`
      [
        {
          "parsedVersion": 8,
          "target": "safari",
          "version": "8",
        },
        {
          "parsedVersion": 9,
          "target": "ie",
          "version": "9",
        },
        {
          "parsedVersion": 20,
          "target": "firefox",
          "version": "20",
        },
        {
          "parsedVersion": 32,
          "target": "chrome",
          "version": "32",
        },
      ]
    `);
  });

  it("should get lowest target versions", () => {
    const versions = [
      "chrome 20",
      "chrome 30",
      "node 7",
      "chrome 30.5",
      "firefox 50.5",
    ];
    expect(parseBrowsersListVersion(versions)).toMatchSnapshot();
  });

  it("should support string config in rule option", () => {
    const config = determineTargetsFromConfig(".", "defaults, not ie < 9");
    const result = parseBrowsersListVersion(config);
    expect(result).toMatchSnapshot();
  });

  it("should fail on incorrect browserslist target version", () => {
    expect(() => {
      determineTargetsFromConfig(".", "edge 100000");
    }).toThrow("Unknown version 100000 of edge");
  });
});
