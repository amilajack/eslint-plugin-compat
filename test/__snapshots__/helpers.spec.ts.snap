// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Versioning should get lowest target versions 1`] = `
[
  {
    "parsedVersion": 7,
    "target": "node",
    "version": "7",
  },
  {
    "parsedVersion": 50.5,
    "target": "firefox",
    "version": "50.5",
  },
  {
    "parsedVersion": 20,
    "target": "chrome",
    "version": "20",
  },
]
`;

exports[`Versioning should support multi env config in browserslist package.json 1`] = `
[
  {
    "parsedVersion": 16,
    "target": "samsung",
    "version": "16.0",
  },
  {
    "parsedVersion": 15.6,
    "target": "safari",
    "version": "15.6",
  },
  {
    "parsedVersion": 91,
    "target": "opera",
    "version": "91",
  },
  {
    "parsedVersion": 11.5,
    "target": "op_mob",
    "version": "11.5",
  },
  {
    "parsedVersion": 0,
    "target": "op_mini",
    "version": "all",
  },
  {
    "parsedVersion": 2.5,
    "target": "kaios",
    "version": "2.5",
  },
  {
    "parsedVersion": 15.6,
    "target": "ios_saf",
    "version": "15.6",
  },
  {
    "parsedVersion": 10,
    "target": "ie_mob",
    "version": "10",
  },
  {
    "parsedVersion": 9,
    "target": "ie",
    "version": "9",
  },
  {
    "parsedVersion": 102,
    "target": "firefox",
    "version": "102",
  },
  {
    "parsedVersion": 106,
    "target": "edge",
    "version": "106",
  },
  {
    "parsedVersion": 106,
    "target": "chrome",
    "version": "106",
  },
  {
    "parsedVersion": 7,
    "target": "bb",
    "version": "7",
  },
  {
    "parsedVersion": 13.18,
    "target": "baidu",
    "version": "13.18",
  },
  {
    "parsedVersion": 109,
    "target": "android",
    "version": "109",
  },
  {
    "parsedVersion": 13.4,
    "target": "and_uc",
    "version": "13.4",
  },
  {
    "parsedVersion": 13.1,
    "target": "and_qq",
    "version": "13.1",
  },
  {
    "parsedVersion": 109,
    "target": "and_ff",
    "version": "109",
  },
  {
    "parsedVersion": 109,
    "target": "and_chr",
    "version": "109",
  },
]
`;

exports[`Versioning should support resolving browserslist config in subdirectory 1`] = `
[
  {
    "parsedVersion": 70,
    "target": "chrome",
    "version": "70",
  },
]
`;

exports[`Versioning should support resolving browserslist config in subdirectory 2`] = `
[
  {
    "parsedVersion": 18,
    "target": "samsung",
    "version": "18.0",
  },
  {
    "parsedVersion": 15.6,
    "target": "safari",
    "version": "15.6",
  },
  {
    "parsedVersion": 93,
    "target": "opera",
    "version": "93",
  },
  {
    "parsedVersion": 73,
    "target": "op_mob",
    "version": "73",
  },
  {
    "parsedVersion": 0,
    "target": "op_mini",
    "version": "all",
  },
  {
    "parsedVersion": 2.5,
    "target": "kaios",
    "version": "2.5",
  },
  {
    "parsedVersion": 14.5,
    "target": "ios_saf",
    "version": "14.5-14.8",
  },
  {
    "parsedVersion": 102,
    "target": "firefox",
    "version": "102",
  },
  {
    "parsedVersion": 107,
    "target": "edge",
    "version": "107",
  },
  {
    "parsedVersion": 107,
    "target": "chrome",
    "version": "107",
  },
  {
    "parsedVersion": 109,
    "target": "android",
    "version": "109",
  },
  {
    "parsedVersion": 13.4,
    "target": "and_uc",
    "version": "13.4",
  },
  {
    "parsedVersion": 13.1,
    "target": "and_qq",
    "version": "13.1",
  },
  {
    "parsedVersion": 109,
    "target": "and_ff",
    "version": "109",
  },
  {
    "parsedVersion": 109,
    "target": "and_chr",
    "version": "109",
  },
]
`;

exports[`Versioning should support single array config in browserslist package.json 1`] = `
[
  {
    "parsedVersion": 16,
    "target": "samsung",
    "version": "16.0",
  },
  {
    "parsedVersion": 15.6,
    "target": "safari",
    "version": "15.6",
  },
  {
    "parsedVersion": 91,
    "target": "opera",
    "version": "91",
  },
  {
    "parsedVersion": 11.5,
    "target": "op_mob",
    "version": "11.5",
  },
  {
    "parsedVersion": 0,
    "target": "op_mini",
    "version": "all",
  },
  {
    "parsedVersion": 2.5,
    "target": "kaios",
    "version": "2.5",
  },
  {
    "parsedVersion": 15.6,
    "target": "ios_saf",
    "version": "15.6",
  },
  {
    "parsedVersion": 10,
    "target": "ie_mob",
    "version": "10",
  },
  {
    "parsedVersion": 9,
    "target": "ie",
    "version": "9",
  },
  {
    "parsedVersion": 102,
    "target": "firefox",
    "version": "102",
  },
  {
    "parsedVersion": 106,
    "target": "edge",
    "version": "106",
  },
  {
    "parsedVersion": 100,
    "target": "chrome",
    "version": "100",
  },
  {
    "parsedVersion": 7,
    "target": "bb",
    "version": "7",
  },
  {
    "parsedVersion": 13.18,
    "target": "baidu",
    "version": "13.18",
  },
  {
    "parsedVersion": 109,
    "target": "android",
    "version": "109",
  },
  {
    "parsedVersion": 13.4,
    "target": "and_uc",
    "version": "13.4",
  },
  {
    "parsedVersion": 13.1,
    "target": "and_qq",
    "version": "13.1",
  },
  {
    "parsedVersion": 109,
    "target": "and_ff",
    "version": "109",
  },
  {
    "parsedVersion": 109,
    "target": "and_chr",
    "version": "109",
  },
]
`;

exports[`Versioning should support string config in rule option 1`] = `
[
  {
    "parsedVersion": 18,
    "target": "samsung",
    "version": "18.0",
  },
  {
    "parsedVersion": 15.6,
    "target": "safari",
    "version": "15.6",
  },
  {
    "parsedVersion": 93,
    "target": "opera",
    "version": "93",
  },
  {
    "parsedVersion": 73,
    "target": "op_mob",
    "version": "73",
  },
  {
    "parsedVersion": 0,
    "target": "op_mini",
    "version": "all",
  },
  {
    "parsedVersion": 2.5,
    "target": "kaios",
    "version": "2.5",
  },
  {
    "parsedVersion": 14.5,
    "target": "ios_saf",
    "version": "14.5-14.8",
  },
  {
    "parsedVersion": 102,
    "target": "firefox",
    "version": "102",
  },
  {
    "parsedVersion": 107,
    "target": "edge",
    "version": "107",
  },
  {
    "parsedVersion": 107,
    "target": "chrome",
    "version": "107",
  },
  {
    "parsedVersion": 109,
    "target": "android",
    "version": "109",
  },
  {
    "parsedVersion": 13.4,
    "target": "and_uc",
    "version": "13.4",
  },
  {
    "parsedVersion": 13.1,
    "target": "and_qq",
    "version": "13.1",
  },
  {
    "parsedVersion": 109,
    "target": "and_ff",
    "version": "109",
  },
  {
    "parsedVersion": 109,
    "target": "and_chr",
    "version": "109",
  },
]
`;
